### ISSUE TITLE
fsync only called on segment closing

### ISSUE THREAD

**Issue by brancz** [prometheus/tsdb#520]
I was reading the ingestion path of tsdb and was talking to @gouthamve as I think `fsync` is only called on closing a segment, which currently only seems to happen when a segment is "[full](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L482-L486)". A segment is [128MB](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L38) by default, so on unclean shutdown, I believe we may loose up to the most recent 128MB of data.

It looks like https://github.com/prometheus/tsdb/pull/332 removed the WAL being flushed at an interval, being part of the strategy to make:

> It is safe for concurrent reads across different processes.

@gouthamve and I were thinking that instead of fsync'ing at an interval, we fsync in [`flushPage`](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L403) or directly at the end of [`log`](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L481).

First of all, let's validate this and make sure we didn't miss something :slightly_smiling_face: . Otherwise let's make sure we fix this. (I'm actually on vacation, just happened to be reading the code for work unrelated reasons, so it's unlikely I will work on it this week, I'll be back next week)

@gouthamve @brian-brazil @codesome @cstyan @fabxc

**Comment by knweiss** [comment]
FWIW:

1. https://lwn.net/Articles/752063/

> “In short, PostgreSQL assumes that a successful call to fsync() indicates that all data written since the last successful call made it safely to persistent storage. But that is not what the kernel actually does. When a buffered I/O write fails due to a hardware-level error, filesystems will respond differently, but that behavior usually includes discarding the data in the affected pages and marking them as being clean. So a read of the blocks that were just written will likely return something other than the data that was written.“

2. https://danluu.com/file-consistency/

**Comment by brancz** [comment]
I'm aware of the problems with fsync, but not calling it at all is definitely worse :slightly_smiling_face: .

**Comment by fabxc** [comment]
> @gouthamve and I were thinking that instead of fsync'ing at an interval, we fsync in [`flushPage`](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L403) or directly at the end of [`log`](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L481).

fsync is very expensive. Doing it per record or page will likely completely screw performance. Not even SQL databases do this in a normal configuration. For us, sustaining ingestion is more important than persistence (alerts).

**Comment by fabxc** [comment]
Yes, periodic background fdatasync should be sensible. Not sure why it
didn't make it into the new WAL.

**Comment by brancz** [comment]
That makes sense. Do you think periodic sync fsync/fdatasync could mess with live reading the write ahead log? If not, then I imagine (re-)introducing it in the new write ahead log shouldn't have too much of a performance penalty (of course has to be benchmarked). I agree that we don't haven't provide perfect durability, but 128Mb flushing seems a little large.

**Comment by brian-brazil** [comment]
If we're doing that, why not rely on the kernel? It'll be doing it every 60s-ish by default.

**Comment by brancz** [comment]
I wasn't quite aware of the specifics of the linux kernel behavior in this regard, so did some quick research and found that `dirty_expire_centisecs` defines this, which defaults to every 30 seconds on most distros. I think I'd prefer though if we can say something along the lines of "durability is roughly on scrapes by default", in which case I would choose what we previously had as the default: 10 seconds (plus we could be more sure about this across platforms and kernel settings independent).

**Comment by krasi-georgiev** [comment]
> A segment is [128MB](https://github.com/prometheus/tsdb/blob/3929359302170bd79567f9808bafcb7fdf116acf/wal/wal.go#L38) by default, so on unclean shutdown, I believe we may loose up to the most recent 128MB of data.

is this really true? my understanding (after a google crash course) is that the kernel does periodic fsync()(writebacks) in the background so no big data loss should be expected.

> Do you think periodic async fsync/fdatasync could mess with live reading the write ahead log?

My understanding after reading [this](https://www.thomas-krenn.com/en/wiki/Linux_Page_Cache_Basics) is that Once the page is flushed to disk it is cleared from the cache so any sequential reads will cause a new cache to be loaded from the disk which will probably slow down the live wal read.

@brancz have there been any reports for big data loss from the WAL file? 

to me it sounds that unless we have some real big data loss reports we should leave all this to the kernel.

**Comment by krasi-georgiev** [comment]
also this is a nice read on the topic https://medium.com/databasss/on-disk-io-part-1-flavours-of-io-8e1ace1de017

**Comment by brancz** [comment]
There haven't been reports of large dataloss, I was just reading the code and noticed this, and recalled that we used to do fsync on an interval. Your findings seem to match with the comments before yours. I still think the interval should be configured by Prometheus as just by a quick poll with people working around me there are different kernel settings between each Linux distro even, plus we can't monitor the fsync latency if we don't perform it, this seems like a rather important metric to have :slightly_smiling_face: .

**Comment by redbaron** [comment]
> once the page is flushed to disk it is cleared from the cache

This is not how it works at least on Linux. flushing page to disk marks it clean so it can be evicted from memory when system is under memory pressure, but it fsync itself  doesn't force eviction

**Comment by krasi-georgiev** [comment]
@redbaron 	thanks for the clarification. This means that regular fsync shouldn't affect the live WAL reading.

btw @codesome 	just posted in the irc channel an article that is quite interesting.
https://medium.com/@valyala/wal-usage-looks-broken-in-modern-time-series-databases-b62a627ab704
